java-client/src/main/java/com/cumulocity/sdk/client/notification/SubscriberImpl.java

```diff
@@ -573,13 +589,66 @@ public boolean equals(Object o) {
 
-            if (id != null ? !id.equals(that.id) : that.id != null) return false;
-            if (listener != null ? !listener.equals(that.listener) : that.listener != null) return false;
-            return subscribeOperationListener != null ? subscribeOperationListener
-                    .equals(that.subscribeOperationListener) : that.subscribeOperationListener == null;
+            if (!Objects.equals(id, that.id)) return false;
+            if (!Objects.equals(listener, that.listener)) return false;
+            return Objects.equals(subscribeOperationListener, that.subscribeOperationListener);
         }
+    }
 
+    private final class SubscriptionWatcher {
+        private final ScheduledExecutorService executorService;
 
-    }
+        public SubscriptionWatcher() {
+            ThreadFactory threadFactory = new ThreadFactoryBuilder()
+                    .setNameFormat("CumulocitySubscriptionWatcher-scheduler-%d")
+                    .build();
+            executorService = Executors.newSingleThreadScheduledExecutor(threadFactory);
+        }
 
+        public void start() {
+            executorService.scheduleWithFixedDelay(
+                    checkSubscriptionsRunner(),
+                    SUBSCRIPTION_WATCHER_DELAY,
+                    SUBSCRIPTION_WATCHER_DELAY,
+                    TimeUnit.SECONDS);
+        }
+
+        public void stop() {
+            executorService.shutdown();
+        }
+
+        private Runnable checkSubscriptionsRunner() {
+            return () -> {
+                log.debug("Running watcher to check subscriptions");
+                ImmutableSet.<SubscriptionRecord>builder()
+                        .addAll(subscriptions)
+                        .addAll(pendingSubscriptions)
+                        .build()
+                        .forEach(this::checkSubscription);
+            };
+        }
+
+        private void checkSubscription(SubscriptionRecord subRec) {
+            ClientSessionChannel channel = getChannel(subRec.getId());
+            if (channel.getSubscribers().isEmpty()) {
+                log.warn("{} bayeux channel {} has no client subscribers",
+                        pendingSubscriptions.contains(subRec) ? "Pending" : "Subscribed", subRec.getId());
+                reSubscribe(subRec);
+            } else {
+                log.debug("Bayeux channel {} has {} client subscriptions (OK)",
+                        subRec.getId(), channel.getSubscribers().size());
+            }
+        }
+
+        private void reSubscribe(SubscriptionRecord subRec) {
+            if (session.isConnected()) {
+                log.info("Trying to subscribe channel {}", subRec.getId());
+                subscribe(subRec.getId(), subRec.subscribeOperationListener, subRec.listener, true);
+            } else {
+                log.warn("Session is not connected, adding channel {} to pending subscriptions", subRec.getId());
+                subscriptions.remove(subRec);
```

Comment:
The non-atomicity of removing from subscriptions to pendingSubscriptions worries me a little, but it will only be an issue if reSubscribe(...) can be called concurrently with resubscribeAll(), (from line 186 approx) or any other method that copies both Sets into one Set. 
Also the SubscriptionWatcher may eventually resolve the issue through it normal, repeated, operation.

The atomicity concern is: if subscriptions.remove(subRec) is called here, and then this thread is paused (context-switched off a core) then the 2 Sets are copied into 1 elsewhere (missing subRec) and then when this thread wakes up subRec is added to pendingSubscriptions. That's okay from this methods point of view, but maybe not from the other thread IF that thread that is expecting to handle both subscriptions and pendingSubscriptions elements including subRec.

This might not be possible - I'm looking at this stuff for the first time and might need to better understand the wider scope to know if these calls can occur concurrently and even if missing subRec is a problem.

The only ways to avoid atomicity issues caused by moving stuff from one collection to another is to wrap them in a lock that is also used in the conflicting places (as I'm 100% sure you know), or to leave them in one collection and give them an atomic pending flag. That might be single primitive boolean if single-writer/many-readers pattern can be used or an AtomicBoolean otherwise.

Like I say, I'm not sure this is an issue so won't request a change. I might continue looking at this, but really probably need a few days to understand it properly as I know nothing about Bayeux and realtime

Reply:
@sag-tjegorow, I suspect this isn't an issue btw. I just can't _yet_ see that it is not

Reply:
We discussed this yesterday and we agreed on our understanding of this. I am happy as it is (don't let perfection stand in the say of improvement) or if ether of the other options are taken (locking the places where both sets are mutated against each other or changing the order of the removes as sets will de-duplicate for us).

I'll resolve this comment and approve. If changes are made let me know if I need to re-approve

Reply:
We have discussed this but I made some improvements in this area (as also discussed).